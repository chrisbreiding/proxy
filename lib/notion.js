const rp = require('request-promise')

const getData = ({ notionPageId, notionToken }) => {
  return rp({
    uri: `https://api.notion.com/v1/blocks/${notionPageId}/children`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

module.exports = {
  getData,
}
