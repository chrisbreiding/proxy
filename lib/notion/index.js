const { addUpcomingWeek } = require('./upcoming-week')
const { getData } = require('./data')
const { onSocket } = require('./shopping')

module.exports = {
  addUpcomingWeek,
  getData,
  onSocket,
}
