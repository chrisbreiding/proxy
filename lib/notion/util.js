const rp = require('request-promise')

const makeRequest = ({ notionToken, body, method = 'get', path }) => {
  return rp({
    method,
    body,
    uri: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

const getBlockChildren = ({ notionToken, pageId }) => {
  return makeRequest({ notionToken, path: `blocks/${pageId}/children` })
}

const appendBlockChildren = ({ notionToken, pageId, blocks }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
}

const updateBlock = ({ notionToken, blockId, block }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: block,
  })
}

const getPlainText = (withText) => {
  if (!withText || !withText.text) return

  return withText.text
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

const weatherIconMap = {
  'clear-day': '☀️',
  'clear-night': '☀️',
  'rain': '☔️',
  'snow': '❄️',
  'sleet': '🌨',
  'wind': '💨',
  'fog': '🌫',
  'cloudy': '☁️',
  'partly-cloudy-day': '⛅️',
  'partly-cloudy-night': '⛅️',
  'default': '🌑',
}

const getWeatherIcon = (iconName) => {
  return weatherIconMap[iconName] || weatherIconMap.default
}

const makeTextPart = (content, color) => {
  const textPart = { text: { content } }

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

module.exports = {
  getBlockChildren,
  appendBlockChildren,
  updateBlock,
  getPlainText,
  getWeatherIcon,
  makeTextPart,
}
