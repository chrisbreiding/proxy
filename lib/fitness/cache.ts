import deepEqual from 'deep-equal'

import type { CacheValue, ChallengeDetails, MMFUserId } from './types'
import dayjs from 'dayjs'
import { PersistentData } from '../util/persistent-data'

export interface PersistentDataStructure {
  cache: {
    [key: string]: CacheValue
  }
}

const persistentData = new PersistentData<PersistentDataStructure>('fitness-data')

async function getCache (userId: MMFUserId) {
  const data = await persistentData.get()

  return (data?.cache || {})[userId]
}

async function setCache (userId: MMFUserId, value: CacheValue) {
  const data = (await persistentData.get()) || { cache: {} }

  data.cache[userId] = value

  await persistentData.set(data)
}

function isCachedValueValid (latestChallengeDetails: ChallengeDetails, value?: CacheValue) {
  return (
    !!value
    && value.date === dayjs().format('YYYY-MM-DD')
    && deepEqual(value.details, latestChallengeDetails)
  )
}

export async function areChallengeDetailsUnchangedToday (userId: MMFUserId, latestChallengeDetails: ChallengeDetails) {
  const cachedValue = await getCache(userId)

  return isCachedValueValid(latestChallengeDetails, cachedValue)
}

export async function setCachedValue (userId: MMFUserId, value: ChallengeDetails) {
  await setCache(userId, {
    date: dayjs().format('YYYY-MM-DD'),
    details: value,
  })
}
