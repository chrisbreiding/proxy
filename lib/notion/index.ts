import { addUpcomingWeek, upcomingWeekView } from './upcoming-week'
import { getBlockChildren } from './util'
import { onSocket } from './shopping'

interface GetDataOptions {
  notionToken: string
  notionPageId: string
}

export function getNotionData ({ notionToken, notionPageId }: GetDataOptions) {
  return getBlockChildren({ notionToken, pageId: notionPageId })
}

export {
  addUpcomingWeek,
  upcomingWeekView,
  onSocket,
}
