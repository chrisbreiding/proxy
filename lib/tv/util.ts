import type { Method } from 'axios'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { request } from '../util/network'

const apikey = getEnv('THETVDB_API_KEY')
const pin = getEnv('THETVDB_PIN')
export const baseUrl = 'https://api4.thetvdb.com'

interface LoginResult {
  status: string
  data: {
    token: string
  }
}

async function authenticate () {
  debugVerbose('authenticate')

  try {
    const { data, status } = await request({
      method: 'post',
      url: `${baseUrl}/v4/login`,
      body: { apikey, pin },
    }) as LoginResult

    if (!data?.token) {
      throw new Error(`Could not authenticate with TheTVDB, status: ${status}`)
    }

    return data.token as string
  } catch (error: any) {
    debug('Authentication failed:', error?.stack || error)

    throw error
  }
}

interface MakeRequestOptions {
  method?: Method
  params?: { [key: string]: string }
  path: string
  token?: string
}

export async function makeRequest (options: MakeRequestOptions) {
  const { method, params, path } = options
  const token = options.token || await authenticate()
  const url = `${baseUrl}/v4/${path}`

  debugVerbose('request: %o', { method, url, params, token })

  // TODO: handle error
  const result = await request({
    headers: {
      'Authorization': `Bearer ${token}`,
    },
    method,
    params,
    url,
  })

  result.token = token

  return result
}
