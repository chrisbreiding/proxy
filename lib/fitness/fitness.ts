import dayjs, { Dayjs } from 'dayjs'
import duration from 'dayjs/plugin/duration'

import { makeBlock, makeDivider, makeRichText } from '../notion/util/general'
import { getBlockChildren } from '../notion/util/queries'
import { appendBlockChildren, deleteBlock } from '../notion/util/updates'
import { compact } from '../util/collections'
import { debug } from '../util/debug'
import { fetchChallengeDetails } from './mmf'
import type { Breakdown, ChallengeDetails, Stats } from './types'
import { areChallengeDetailsUnchangedToday, setCachedValue } from './cache'

dayjs.extend(duration)

function getDates () {
  const now = dayjs()
  const startOfYear = dayjs(`${now.year()}-01-01`)
  const daysPassedIncludingToday = Math.ceil((now.diff(startOfYear, 'day')))
  const daysLeftThroughWeekendExcludingToday = now.day() === 0 ? 0 : 7 - now.day() // through Sunday
  const daysPassedAtEndOfWeek = daysPassedIncludingToday + daysLeftThroughWeekendExcludingToday

  return {
    daysLeftThroughWeekendExcludingToday,
    daysPassedAtEndOfWeek,
    daysPassedIncludingToday,
    now,
  }
}

function getBreakdown (progress: ChallengeDetails['progress'], dates: ReturnType<typeof getDates>) {
  const { daysLeftThroughWeekendExcludingToday, daysPassedAtEndOfWeek, daysPassedIncludingToday } = dates

  const yearlyGoal = (progress.goal / 1000) * 0.621371 // miles
  const dailyGoal = yearlyGoal / 365
  const totalProgressMade = (progress.currentScore / 1000) * 0.621371 // miles
  const yearlyPercentComplete = totalProgressMade / yearlyGoal * 100

  const untilEndOfTodayGoal = dailyGoal * daysPassedIncludingToday
  const untilEndOfTodayDifference = totalProgressMade - untilEndOfTodayGoal
  const aheadOfGoalForToday = untilEndOfTodayDifference >= 0
  const milesNeededToBeOnTrackToday = aheadOfGoalForToday ? 0 : Math.abs(untilEndOfTodayDifference)

  const untilEndOfWeekendGoal = daysPassedAtEndOfWeek * dailyGoal
  const untilEndOfWeekendDifference = totalProgressMade - untilEndOfWeekendGoal
  const aheadOfGoalForWeek = untilEndOfWeekendDifference >= 0

  const milesPerDayNeededToBeOnTrackThroughWeekend = (() => {
    if (daysLeftThroughWeekendExcludingToday === 0) return 0
    if (aheadOfGoalForWeek) return 0

    return Math.abs(untilEndOfWeekendDifference) / daysLeftThroughWeekendExcludingToday
  })()

  return {
    aheadOfGoalForToday,
    aheadOfGoalForWeek,
    daysLeftThroughWeekendExcludingToday,
    milesNeededToBeOnTrackToday,
    milesPerDayNeededToBeOnTrackThroughWeekend,
    totalProgressMade,
    untilEndOfTodayDifference,
    untilEndOfWeekendDifference: Math.abs(untilEndOfWeekendDifference),
    yearlyGoal,
    yearlyPercentComplete,
  }
}

function getStats (stats: ChallengeDetails['stats']) {
  const caloriesBurned = stats.calories_burned
  const _totalTime = stats.total_time
  const totalTime = _totalTime ? dayjs.duration(_totalTime, 'seconds').format('H:mm:ss') : undefined
  const totalWorkouts = stats.total_workouts

  return {
    caloriesBurned: caloriesBurned ? `${caloriesBurned}` : '<unknown>',
    totalTime: totalTime ? `${totalTime}` : '<unknown>',
    totalWorkouts: totalWorkouts ? `${totalWorkouts}` : '<unknown>',
  }
}

async function getChallengeDetails (token: string, apiKey: string, userId: string, date: Dayjs) {
  const challengeDetails = await fetchChallengeDetails(token, apiKey, userId)

  if (await areChallengeDetailsUnchangedToday(userId, challengeDetails, date)) {
    return {
      changed: false,
      challengeDetails,
    }
  }

  return {
    changed: true,
    challengeDetails,
  }
}

function makeStat (value: string, label: string) {
  return makeBlock({
    content: {
      rich_text: [makeRichText(value, { bold: true }), makeRichText(label)],
    },
    type: 'bulleted_list_item',
  })
}

function t (text: string, bold = false) {
  return makeRichText(text, { bold })
}

type Color = 'default' | 'green' | 'orange' | 'red' | 'blue' | 'gray'

function b (type: 'paragraph' | 'bulleted_list_item', richTexts: ReturnType<typeof t>[], color: Color = 'default') {
  return makeBlock({
    content: { rich_text: richTexts, color },
    type,
  })
}

function p (richTexts: ReturnType<typeof t>[], color: Color = 'default') {
  return b('paragraph', richTexts, color)
}

function li (richTexts: ReturnType<typeof t>[], color: Color = 'default') {
  return b('bulleted_list_item', richTexts, color)
}

async function clearExistingBlocks (token: string, blockId: string) {
  debug('Deleting existing blocks...')

  const existingBlocks = await getBlockChildren({ notionToken: token, pageId: blockId })

  for (const block of existingBlocks) {
    await deleteBlock({ notionToken: token, id: block.id })
  }
}

function pluralize (value: number, label: string) {
  return toTwoDecimals(value) === 1 ? label : `${label}s`
}

function logDashboardToDebug (breakdown: Breakdown, stats: Stats) {
  const {
    aheadOfGoalForToday,
    aheadOfGoalForWeek,
    daysLeftThroughWeekendExcludingToday,
    milesNeededToBeOnTrackToday,
    milesPerDayNeededToBeOnTrackThroughWeekend,
    totalProgressMade,
    untilEndOfTodayDifference,
    untilEndOfWeekendDifference,
    yearlyGoal,
    yearlyPercentComplete,
  } = breakdown
  const { caloriesBurned, totalTime, totalWorkouts } = stats

  const todayDebugStatus = aheadOfGoalForToday ?
    `You're ${toTwoDecimals(untilEndOfTodayDifference)} ${pluralize(untilEndOfTodayDifference, 'mile')} ahead!`
    : `Needed to be on track today
    - ${toTwoDecimals(milesNeededToBeOnTrackToday)} ${pluralize(milesNeededToBeOnTrackToday, 'mile')}`

  const weekDebugStatus = aheadOfGoalForWeek ?
    `You're ${toTwoDecimals(untilEndOfWeekendDifference)} ${pluralize(untilEndOfWeekendDifference, 'mile')} ahead!`
    : `Needed to be on track this week
    - ${toTwoDecimals(untilEndOfWeekendDifference)} ${pluralize(untilEndOfWeekendDifference, 'mile')} / ${daysLeftThroughWeekendExcludingToday} ${pluralize(daysLeftThroughWeekendExcludingToday, 'day')}
    - ${toTwoDecimals(milesPerDayNeededToBeOnTrackThroughWeekend)} ${pluralize(milesPerDayNeededToBeOnTrackThroughWeekend, 'mile')} per day`

  debug(`
    ${toTwoDecimals(yearlyPercentComplete)}% complete
    ${toTwoDecimals(totalProgressMade)} / ${Math.round(yearlyGoal)} miles
    ---
    ${todayDebugStatus}
    ${weekDebugStatus}
    ---
    Stats
    - ${caloriesBurned} calories burned
    - ${totalTime} total time
    - ${totalWorkouts} total workouts
  `)
}

function getTodayStatus (breakdown: Breakdown) {
  const { aheadOfGoalForToday, milesNeededToBeOnTrackToday, untilEndOfTodayDifference } = breakdown

  if (aheadOfGoalForToday) {
    return [
      p([
        t('You\'re '),
        t(`${toTwoDecimals(untilEndOfTodayDifference)}`, true),
        t(` ${pluralize(untilEndOfTodayDifference, 'mile')} ahead for today!`),
      ], 'green'),
    ]
  }

  return [
    p([t('Needed to be on track today')]),
    li([
      t(`${toTwoDecimals(milesNeededToBeOnTrackToday)}`, true),
      t(` ${pluralize(milesNeededToBeOnTrackToday, 'mile')}`),
    ]),
  ]
}

function getWeekStatus (breakdown: Breakdown) {
  const {
    aheadOfGoalForWeek,
    milesNeededToBeOnTrackToday,
    untilEndOfWeekendDifference,
    daysLeftThroughWeekendExcludingToday,
    milesPerDayNeededToBeOnTrackThroughWeekend,
  } = breakdown

  if (aheadOfGoalForWeek) {
    return [
      p([
        t('You\'re '),
        t(`${toTwoDecimals(untilEndOfWeekendDifference)}`, true),
        t(` ${pluralize(untilEndOfWeekendDifference, 'mile')} ahead for the week!`),
      ], 'green'),
    ]
  }

  if (daysLeftThroughWeekendExcludingToday === 0) {
    return [
      p([t('Needed to be on track this week')]),
      li([
        t(`${toTwoDecimals(milesNeededToBeOnTrackToday)}`, true),
        t(` ${pluralize(milesNeededToBeOnTrackToday, 'mile')} today`),
      ]),
    ]
  }

  return [
    p([t('Needed to be on track this week')]),
    li([
      t(`${toTwoDecimals(untilEndOfWeekendDifference)}`, true),
      t(` ${pluralize(untilEndOfWeekendDifference, 'mile')} / ${daysLeftThroughWeekendExcludingToday} ${pluralize(daysLeftThroughWeekendExcludingToday, 'day')}`),
    ]),
    li([
      t(`${toTwoDecimals(milesPerDayNeededToBeOnTrackThroughWeekend)}`, true),
      t(` ${pluralize(milesPerDayNeededToBeOnTrackThroughWeekend, 'mile')} per day`),
    ]),
  ]
}

async function updateFitnessDashboard (breakdown: Breakdown, stats: Stats, token: string, blockId: string, date: Dayjs) {
  const {
    totalProgressMade,
    yearlyGoal,
    yearlyPercentComplete,
  } = breakdown
  const { caloriesBurned, totalTime, totalWorkouts } = stats

  await clearExistingBlocks(token, blockId)

  logDashboardToDebug(breakdown, stats)

  const blocks = compact([
    p([
      t(`${toTwoDecimals(yearlyPercentComplete)}%`, true),
      t(' complete'),
    ]),
    p([
      t(`${toTwoDecimals(totalProgressMade)}`, true),
      t(` / ${Math.round(yearlyGoal)} miles`),
    ]),
    makeDivider(),
    ...getTodayStatus(breakdown),
    ...getWeekStatus(breakdown),
    makeDivider(),
    p([t('Stats')]),
    makeStat(caloriesBurned, ' calories burned'),
    makeStat(totalTime, ' total time'),
    makeStat(totalWorkouts, ' total workouts'),
    makeDivider(),
    p([t('Updated '), t(`${date.format('MMM D, h:mma')}`)], 'gray'),
  ])

  debug('Adding blocks...')

  await appendBlockChildren({ notionToken: token, pageId: blockId, blocks })
}

function toTwoDecimals (value: number) {
  return Number(value.toFixed(2))
}

interface UpdateFitnessProps {
  mmfToken: string
  mmfApiKey: string
  mmfUserId: string
  notionToken: string
  notionFitnessId: string
}

export async function updateFitness (props: UpdateFitnessProps) {
  const { mmfToken, mmfApiKey, mmfUserId, notionToken, notionFitnessId } = props
  const dates = getDates()
  const { changed, challengeDetails } = await getChallengeDetails(mmfToken, mmfApiKey, mmfUserId, dates.now)

  if (!changed) {
    debug('Challenge details have not changed, skipping update')

    return
  }

  const breakdown = getBreakdown(challengeDetails.progress, dates)
  const stats = getStats(challengeDetails.stats)

  await updateFitnessDashboard(breakdown, stats, notionToken, notionFitnessId, dates.now)
  await setCachedValue(mmfUserId, challengeDetails, dates.now)
}
