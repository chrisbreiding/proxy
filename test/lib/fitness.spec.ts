import mockFs from 'mock-fs'
import nock from 'nock'
import path from 'path'
import { afterAll, afterEach, beforeEach, describe, it, vi } from 'vitest'
import { updateFitness } from '../../lib/fitness/fitness'
import { fixtureContents } from '../util'
import { block, getAppendBody, listResults, nockAppendBlockChildren, nockDeleteBlock, nockGetBlockChildren, snapshotAppendChildren } from './notion/util'
import type { PersistentDataStructure } from '../../lib/fitness/cache'

function mockPersistentData (data: PersistentDataStructure | null) {
  mockFs({
    'data': {
      'fitness-data.json': JSON.stringify(data),
    },
    'test': {
      'fixtures': {
        'fitness': {
          'challenge-details.json': mockFs.load(path.resolve(process.cwd(), 'test/fixtures/fitness/challenge-details.json')),
        },
      },
    },
  })
}

function nockFitness (currentScore?: number, transform?: (fitness: any) => any) {
  let data = fixtureContents('fitness/challenge-details')

  if (currentScore) {
    data.progress.currentScore = currentScore
  }

  if (transform) {
    data = transform(data)
  }

  nock('https://api.mapmyfitness.com')
  .get('/challenges/challenge/yvty2025/details')
  .matchHeader('authorization', 'Bearer mmf-token')
  .matchHeader('apiKey', 'mmf-api-key')
  .query({
    mmfUserId: 'mmf-user-id',
    timezone: 'America/New_York',
  })
  .reply(200, data)
}

function nockAndSnapshotDashboard () {
  nockGetBlockChildren('notion-fitness-id', { reply: { results: [] } })

  return snapshotAppendChildren({ id: 'notion-fitness-id' })
}

function nockDashboard () {
  nockGetBlockChildren('notion-fitness-id', { reply: { results: [] } })

  const scope = nockAppendBlockChildren({ id: 'notion-fitness-id' })

  return getAppendBody(scope)
}

describe('lib/fitness', () => {
  const props = {
    mmfToken: 'mmf-token',
    mmfApiKey: 'mmf-api-key',
    mmfUserId: 'mmf-user-id',
    notionToken: 'notion-token',
    notionFitnessId: 'notion-fitness-id',
  }

  beforeEach(() => {
    vi.useFakeTimers()
    vi.setSystemTime(new Date(2022, 0, 11))
    mockPersistentData(null)
  })

  afterEach(() => {
    vi.useRealTimers()
  })

  afterAll(() => {
    mockFs.restore()
  })

  it('display fitness breakdown and stats', async () => {
    nockFitness()

    const snapshot = nockAndSnapshotDashboard()

    await updateFitness(props)

    await snapshot
  })

  it('shows a different message if on track for today', async () => {
    nockFitness(32000)

    const snapshot = nockAndSnapshotDashboard()

    await updateFitness(props)

    await snapshot
  })

  it('shows a different message if on track for the week', async () => {
    vi.setSystemTime(new Date(2022, 0, 6))
    nockFitness(29000)

    const snapshot = nockAndSnapshotDashboard()

    await updateFitness(props)

    await snapshot
  })

  it('handles Sunday when behind', async () => {
    vi.setSystemTime(new Date(2022, 0, 9))
    nockFitness(20000)

    const snapshot = nockAndSnapshotDashboard()

    await updateFitness(props)

    await snapshot
  })

  it('handles Sunday when ahead', async () => {
    vi.setSystemTime(new Date(2022, 0, 9))
    nockFitness(30000)

    const snapshot = nockAndSnapshotDashboard()

    await updateFitness(props)

    await snapshot
  })

  it('handles pluralization when ahead for today and for the week', async () => {
    vi.setSystemTime(new Date(2022, 0, 9))
    nockFitness(26070)

    const snapshot = nockAndSnapshotDashboard()

    await updateFitness(props)

    await snapshot
  })

  it('handles pluralization when behind for today', async () => {
    vi.setSystemTime(new Date(2022, 0, 8))
    nockFitness(18051)

    const snapshot = nockAndSnapshotDashboard()

    await updateFitness(props)

    await snapshot
  })

  it('handles pluralization when behind for the week', async () => {
    nockFitness(40520)

    const snapshot = nockAndSnapshotDashboard()

    await updateFitness(props)

    await snapshot
  })

  it('handles pluralization when behind for the week per day', async () => {
    nockFitness(34100)

    const snapshot = nockAndSnapshotDashboard()

    await updateFitness(props)

    await snapshot
  })

  it('handles stats not being defined', async () => {
    nockFitness(20000, (data) => {
      data.stats = []
      return data
    })

    const snapshot = nockAndSnapshotDashboard()

    await updateFitness(props)

    await snapshot
  })

  it('does not update notion if the challenge details have not changed and it is the same day', async () => {
    vi.setSystemTime(new Date(2022, 1, 10))
    nockFitness()
    nockDashboard()

    await updateFitness(props)

    vi.setSystemTime(new Date(2022, 1, 10))
    nockFitness()

    // causes nock to throw on any request
    nock.disableNetConnect()

    // nock will throw an error if the request is made, otherwise this will pass
    await updateFitness(props)
  })

  it('updates notion if the challenge details have changed', async () => {
    nockFitness(34100)
    nockDashboard()

    await updateFitness(props)

    nockFitness(34101)
    const requestMade = nockDashboard()

    await updateFitness(props)

    await requestMade
  })

  it('updates notion if it is a new day', async () => {
    vi.setSystemTime(new Date(2022, 1, 20))

    nockFitness()
    nockDashboard()

    await updateFitness(props)

    vi.setSystemTime(new Date(2022, 1, 21))
    nockFitness()
    const requestMade = nockDashboard()

    await updateFitness(props)

    await requestMade
  })

  it('deletes existing blocks', async () => {
    nockFitness(34100)
    nockGetBlockChildren('notion-fitness-id', { reply: listResults([
      block.p({ id: 'existing-block-id', text: 'existing block' }),
    ]) })
    nockDeleteBlock('existing-block-id')
    nockAppendBlockChildren({ id: 'notion-fitness-id' })

    await updateFitness(props)
  })
})
