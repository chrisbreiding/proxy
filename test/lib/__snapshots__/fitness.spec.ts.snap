// Vitest Snapshot v1

exports[`lib/fitness > display fitness breakdown and stats 1`] = `
"**2.24%** complete
**14.29** / 637 miles
---
Needed to be on track today
- **4.9** miles
Needed to be on track this week
- **13.63** miles / 5 days
- **2.73** miles per day
---
Stats
- **2138** calories burned
- **5:11:25** total time
- **8** total workouts
---
Updated Jan 11, 12:00am
"
`;

exports[`lib/fitness > handles Sunday when ahead 1`] = `
"**2.93%** complete
**18.64** / 637 miles
---
You're **2.94** miles ahead for today!
You're **2.94** miles ahead for the week!
---
Stats
- **2138** calories burned
- **5:11:25** total time
- **8** total workouts
---
Updated Jan 9, 12:00am
"
`;

exports[`lib/fitness > handles Sunday when behind 1`] = `
"**1.95%** complete
**12.43** / 637 miles
---
Needed to be on track today
- **3.28** miles
Needed to be on track this week
- **3.28** miles today
---
Stats
- **2138** calories burned
- **5:11:25** total time
- **8** total workouts
---
Updated Jan 9, 12:00am
"
`;

exports[`lib/fitness > handles pluralization when ahead for today and for the week 1`] = `
"**2.54%** complete
**16.2** / 637 miles
---
You're **0.49** miles ahead for today!
You're **0.49** miles ahead for the week!
---
Stats
- **2138** calories burned
- **5:11:25** total time
- **8** total workouts
---
Updated Jan 9, 12:00am
"
`;

exports[`lib/fitness > handles pluralization when behind for the week 1`] = `
"**3.95%** complete
**25.18** / 637 miles
---
You're **5.98** miles ahead for today!
Needed to be on track this week
- **2.74** miles / 5 days
- **0.55** miles per day
---
Stats
- **2138** calories burned
- **5:11:25** total time
- **8** total workouts
---
Updated Jan 11, 12:00am
"
`;

exports[`lib/fitness > handles pluralization when behind for the week per day 1`] = `
"**3.33%** complete
**21.19** / 637 miles
---
You're **1.99** miles ahead for today!
Needed to be on track this week
- **6.73** miles / 5 days
- **1.35** miles per day
---
Stats
- **2138** calories burned
- **5:11:25** total time
- **8** total workouts
---
Updated Jan 11, 12:00am
"
`;

exports[`lib/fitness > handles pluralization when behind for today 1`] = `
"**1.76%** complete
**11.22** / 637 miles
---
Needed to be on track today
- **2.74** miles
Needed to be on track this week
- **4.49** miles / 1 day
- **4.49** miles per day
---
Stats
- **2138** calories burned
- **5:11:25** total time
- **8** total workouts
---
Updated Jan 8, 12:00am
"
`;

exports[`lib/fitness > handles stats not being defined 1`] = `
"**1.95%** complete
**12.43** / 637 miles
---
Needed to be on track today
- **6.77** miles
Needed to be on track this week
- **15.49** miles / 5 days
- **3.1** miles per day
---
Stats
- **<unknown>** calories burned
- **<unknown>** total time
- **<unknown>** total workouts
---
Updated Jan 11, 12:00am
"
`;

exports[`lib/fitness > shows a different message if on track for the week 1`] = `
"**2.83%** complete
**18.02** / 637 miles
---
You're **7.55** miles ahead for today!
You're **2.32** miles ahead for the week!
---
Stats
- **2138** calories burned
- **5:11:25** total time
- **8** total workouts
---
Updated Jan 6, 12:00am
"
`;

exports[`lib/fitness > shows a different message if on track for today 1`] = `
"**3.12%** complete
**19.88** / 637 miles
---
You're **0.69** miles ahead for today!
Needed to be on track this week
- **8.04** miles / 5 days
- **1.61** miles per day
---
Stats
- **2138** calories burned
- **5:11:25** total time
- **8** total workouts
---
Updated Jan 11, 12:00am
"
`;
